dlfcn.c

```diff
@@ -463,6 +465,92 @@ char *dlerror( void )
     return error_buffer;
 }
 
+#ifdef __USE_GNU
+/*
+ * Return adress from Image Allocation Table (iat), if
+ * the original address points to a thunk table entry.
+ */
+static void *getAddressFromIAT( const void *addr )
+{
+    const unsigned char *p = addr;
+    /* ...inline app code...
+     * 00401002  |. E8 7B0D0000    CALL 00401D82               ; \GetModuleHandleA
+     * ...thunk table...
+     * 00401D82   $-FF25 4C204000 , JMP DWORD PTR DS:[40204C]  ;  KERNEL32.GetModuleHandleA
+     * ...memory address value of pointer...
+     * 40204C > FC 3D 57 7C   ;little endian pointer value
+     */
+    if( *p == 0xff && *(p+1) == 0x25 )
+    {
+        void **p1 = (void *)(p+2);
+        void **p2 = *p1;

```

Comment:
This looks very fragile and can lead to segfaults.

I think that dladdr should not crash when its argument is memory address of some buffer.

E.g. calling `dladdr(buffer)` for variable specified by `unsigned char buffer[2] = "\xFF\x25";` cause memory corruption.

Cannot we check if passed address is in IAT? Instead of doing these dereferences also on non-IAT address?

Reply:
For inspiration, here is code for parsing IAT:
https://relearex.wordpress.com/2017/12/26/hooking-series-part-i-import-address-table-hooking/

I think similar thing could be used to convert IAT function pointer to real function pointer and also to function name.

Reply:
> I think similar thing could be used to convert IAT function pointer to real function pointer 

This is already used by the recent implementation. Are you taking about a difference approach ?

> and also to function name.

That will only work for functions, that are not imported by original.  It requires to investigate how to get that symbol name.

Reply:
> > I think similar thing could be used to convert IAT function pointer to real function pointer
> 
> This is already used by the recent implementation.

Seems not as in recent implementation I still see two indirect pointer derefences which may cause segfaults.

> Are you taking about a difference approach ?

Yes, it is described in that blog post. You load and partse IAT table and then find pointer (passed in dladdr function) in parsed IAT table.

> > and also to function name.
> 
> That will only work for functions, that are not imported by original. It requires to investigate how to get that symbol name.

It should work for at least something and symbol name you have in IAT table. It is stored in structure member `Name`.